README.md
| [![Redmine CRM Plugin](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/1_Box_CRM_1.png)](https://secure.avangate.com/order/product.php?PRODS=4545039&QTY=1&AFFILIATE=63933)                      | [![Redmine HelpDesk plugin](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/1_Box_helpdesk_1.png)](https://secure.avangate.com/order/product.php?PRODS=4549016&QTY=1&AFFILIATE=63933)  | [![Redmine Invoices plugin](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/Box_invoices1.png)](https://secure.avangate.com/order/product.php?PRODS=4550948&QTY=1&AFFILIATE=63933)     | [![Redmine Agile plugin](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/copy_Box_agile.png)](https://secure.avangate.com/order/product.php?PRODS=4619337&QTY=1&AFFILIATE=63933)   |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| [![Redmine Finance plugin](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/1_Box_finance.png)](https://secure.avangate.com/order/product.php?PRODS=4581548&QTY=1&AFFILIATE=63933)                | [![Redmine People plugin](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/Box_people.png)](https://secure.avangate.com/order/product.php?PRODS=4667353&QTY=1&AFFILIATE=63933)          | [![Redmine Products plugin](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/copy_Box_products.png)](https://secure.avangate.com/order/product.php?PRODS=4614495&QTY=1&AFFILIATE=63933) | [![Redmine MailChimp plugin](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/mailchimp.png)](https://secure.avangate.com/order/product.php?PRODS=4637012&QTY=1&AFFILIATE=63933)    |
| [![Redmine Checklists plugin](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/Box_checklists.png)](https://secure.avangate.com/order/product.php?PRODS=4642832&QTY=1&AFFILIATE=63933)            | [![CRM + Helpdesk bundle](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/Box_helpdesk_bundle.png)](https://secure.avangate.com/order/product.php?PRODS=4609660&QTY=1&AFFILIATE=63933) | [![CRM + Helpdesk + Invoices bundle](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/3_Bundle.png)](https://secure.avangate.com/order/product.php?PRODS=4609671&QTY=1&AFFILIATE=63933) | [![RedmineUP full stack bundle](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/copy_logo.png)](https://secure.avangate.com/order/product.php?PRODS=4614499&QTY=1&AFFILIATE=63933) |
| [![Redmine Cloud - Monthly/Annual Subscription](https://secure.avangate.com/images/merchant/ef452c63f81d0105dd4486f775adec81/products/hosting.png)](https://secure.avangate.com/order/product.php?PRODS=4682493&QTY=1&AFFILIATE=63933) | [![Visit Website](https://www.redmineup.com/cms/assets/download/33292/logo_color_up.svg)](https://secure.avangate.com/affiliate.php?ACCOUNT=KIRILLBE&AFFILIATE=63933&PATH=http%3A%2F%2Fredmineup.com)                        |                                                                                                                                                                                                                              |                                                                                                                                                                                                                          |


doc/CHANGELOG
== Redmine CRM plugin changelog

Redmine CRM plugin - customer relationship management plugin for redmine
Copyright (C) 2010-2018 RedmineUP
http://www.redmineup.com/

== 2018-02-22 v4.1.2

* Redmine 4.0 dev support
* Select2 control issue's deal  
* New CSV export UI for Contacts
* Deals layout fixes
* Show contacts deals on a company profile
* Avatar image on click action changed to show 
* Fixed contacts column output for issues table
* Fixed deals pipeline bugs
* Fixed bug with import_settings
* Added missing assigned_to to Excel export
* Email validator fixes
* Duplications seach fixes

== 2017-07-07 v4.1.1

* Redmine 3.4 support
* New styles for monochrome tags
* Fixed company rename bug
* Email validation fix
* Select2 drop box fix

== 2017-05-30 v4.1.0

* Added Select2 to contacts fields
* Primary email was added to dublicate search 
* Chinese translation update
* Fixed deal price precision bug
* Fixed avatar upload on modal form
* Fixed multivalues field import bug

== 2017-03-02 v4.0.8

* Fixed migration requirements with new redmine_crm gem

== 2017-03-02 v4.0.7

* Select2 control for selecting contacts and companies
* Modal for adding contacts to deal
* Fixed history log for done issue from contact page
* Fixed delete confirmation message for contacts, deals and notes 
* Serbian translations (Aleksandar Pavic)

== 2017-02-07 v4.0.6

* Deal lines with Products plugin
* Email incorrect format validation
* Fixed bulk edit tag fields

== 2016-11-17 v4.0.5

* "+" button support
* German translation update (Marcel Müller)
* New pagination styles
* IDs filters for contacts and deals
* User activity duplication bug fixed
* Sales funnel bug fixed
* Fixed bug with select all contacts
* Fixed russian "б" in tags field
* Fixed enconding bug in vCard export
* Fixed bug with deal amount delimiter
* Fixed bug with sending attaches via email
* Import views compatibility fixes

== 2015-12-02 v4.0.4

* Fixed major currencies bug
* Fixed bug with deals statistics
* Fixed bug with last notes duplication

== 2015-11-06 v4.0.3

* New import UI for redmine 3.2+
* Fixed company contacts show bug
* Fixed notes custom field bug
* Fixed bug with contacts queries
* Fixed bug with bigtext in contact notes
* Fixed custom field bug in Windows Edge
* Fixed bug with deal watchers delete link
* Fixed bug with contact url on contacts list view

== 2015-08-24 v4.0.2

* Recalculate deals amount on board view
* Fixed bug with import State and Country fields
* Fixed bug with search in SQL Server

== 2015-06-18 v4.0.1

* Redmine 3.1 support
* Fixed but with contact avatar preview
* Fixed bug on user creation
* Fixed CSV export error for trunk Redmine version

== 2015-06-15 v4.0.0

* Redmine 3 support
* Spanish Translation update (Luis Blasco)
* Chinese translation (zhoutt)
* Fixed bug with sorting contacts by assignee
* ID field for contacts list

== 2015-03-19 v3.4.5

* List optional filter for contact on issues list
* French translation update (Olivier Houdas)
* Swedish translation update (Khedron Wilk)
* Fixed recursion in contact creation
* redirect_on_sucess param for REST API contact creation
* Portuguese Brazilian Translation update (Leandro Gehlen)
* Spanish Translation update (Luis Blasco)
* Deal probability filter added (Jacek Dyło)
* Fixed bug with deal category deletion (Jacek Dyło)

== 2014-11-14 v3.4.4

* Custom fields fixes for Redmine < 2.5
* REST API for managing contact avatars
* Deals pipeline list view

== 2014-11-10 v3.4.3

* Deals calendar view
* Company custom field
* Permissions for managing contact issue relations
* Expected revenue field and totals for deals table
* Serbian translation (Radenko)
* Fixed bug with create contact select (Modal window)
* Fixed bug with contacts table columns after quick search

== 2014-10-16 v3.4.2

* Deal custom fields styles
* Contact name formats
* Global search by custom values
* Autocomplete contacts and deals search by tokens
* Deal add contact select limited by 50 contacts
* Cross-project deals autocomplete on issue edit form
* Fixed bug with group by custom field
* Notifications fixes

== 2014-10-02 v3.4.1

* Seaching contacts by email in global redmine search
* Default country for address selection
* Greek translation (Filippos Karapetis)
* CSV import utf-8 fixes
* Fixed global list assignees and authors filters
* Fixed bug with contacts list custom field sorting
* Add uniqueness for contact first name on scope last name, company and email

== 2014-09-21 v3.4.0

* Deal issues
* Spent time reports for deal and deal's contact
* Added currency and background filters for deals
* Added currency field for deals table
* Added formating for probability field
* Permissions check for issue list contact columns and filters
* Fixed bug with cached tags on tags merge, edit and delete
* REST API for contact projects create/destroy
* REST API for contact tag list

== 2014-09-15 v3.3.0

* Deal queries
* REST API for deal statuses, contacts queries and deal categories
* CSV export custsom field columns order fixes (Sascha Hübner)
* CSV deals export empty price fixes
* Added deal name field to REST API
* Fixed bug with user creation from contact page
* Mark Deals menu item as selected on import

== 2014-05-16 v3.2.17

* Deal macro deleted from light version
* New styles for color picker
* Compatibility fixes with other RedmineCRM plugins

== 2014-05-03 v3.2.16

* Permission for export contacs & deals
* French translation update by Olivier Houdas
* Spanish tranlsation update by Leandro Russo
* Merge contacts fixes

== 2014-03-31 v3.2.15

* Translation fixes for notes global search
* Disable unique validation by contact name
* Ajax contact tabs
* Fixed bug with Liquid gem requirements
* Fixed global search by contact note

== 2014-02-25 v3.2.14

* Update Polish translation (Szymon Anders)
* Fixed bug with custom fields grouping in Contacts queries
* Fixed bug with bulk edit contacts with multiple value custom fields
* Fixed bug with date filters in contacts list


assets/javascripts/contacts.js
// Copyright (c) 2011-2013 Kirill Bezrukov

var noteFileFieldCount = 1;

function addNoteFileField() {
	if (noteFileFieldCount >= 10) return false
	noteFileFieldCount++;
	var f = document.createElement("input");
	f.type = "file";
	f.name = "note_attachments[" + noteFileFieldCount + "][file]";
	f.size = 30;
	var d = document.createElement("input");
	d.type = "text";
	d.name = "note_attachments[" + noteFileFieldCount + "][description]";
	d.size = 60;

	p = document.getElementById("note_attachments_fields");
	p.appendChild(document.createElement("br"));
	p.appendChild(f);
	p.appendChild(d);
}

function updateCustomForm(url, form) {
  $.ajax({
    url: url,
    type: 'post',
    data: $(form).serialize()
  });
}

function toggleContact(event, element)
{
	if (event.shiftKey==1)
	{
		if (element.checked) {
			checkAllContacts($$('.contacts.index td.checkbox input'));
		}
		else
		{
			uncheckAllContacts($$('.contacts.index td.checkbox input'));
		}
	}
	else
	{
		Element.up(element, 'tr').toggleClassName('context-menu-selection');
	}
}


// Observ field function

(function( $ ){

  jQuery.fn.observe_field = function(frequency, callback) {

    frequency = frequency * 100; // translate to milliseconds

    return this.each(function(){
      var $this = $(this);
      var prev = $this.val();

      var check = function() {
        if(removed()){ // if removed clear the interval and don't fire the callback
          if(ti) clearInterval(ti);
          return;
        }

        var val = $this.val();
        if(prev != val){
          prev = val;
          $this.map(callback); // invokes the callback on $this
        }
      };

      var removed = function() {
        return $this.closest('html').length == 0
      };

      var reset = function() {
        if(ti){
          clearInterval(ti);
          ti = setInterval(check, frequency);
        }
      };

      check();
      var ti = setInterval(check, frequency); // invoke check periodically

      // reset counter after user interaction
      $this.bind('keyup click mousemove', reset); //mousemove is for selects
    });

  };

  $.fn.insertAtCaret = function (myValue) {

    return this.each(function() {

      //IE support
      if (document.selection) {

        this.focus();
        sel = document.selection.createRange();
        sel.text = myValue;
        this.focus();

      } else if (this.selectionStart || this.selectionStart == '0') {

        //MOZILLA / NETSCAPE support
        var startPos = this.selectionStart;
        var endPos = this.selectionEnd;
        var scrollTop = this.scrollTop;
        this.value = this.value.substring(0, startPos)+ myValue+ this.value.substring(endPos,this.value.length);
        this.focus();
        this.selectionStart = startPos + myValue.length;
        this.selectionEnd = startPos + myValue.length;
        this.scrollTop = scrollTop;

      } else {

        this.value += myValue;
        this.focus();
      }
    });
  };


})( jQuery );

function setupDeferredTabs(url) {
    $('body').on('click', '.tab-header', function(e){
        tab = $(e.target);
        $('.tab-placeholder').removeClass('active');
        name = tab.data('name');
        partial = tab.data('partial');
        placeholder = $('#tab-placeholder-' + name);
        placeholder.addClass('active');

        if (!placeholder.is('.loaded')) {
            url = url
            $.ajax(url, {
                data: {tab_name: name, partial: partial},
                complete: function(){
                    placeholder.addClass('loaded')
                    //replaces current URL with the "href" attribute of the current link
                    //(only triggered if supported by browser)
                    if ("replaceState" in window.history) {
                      window.history.replaceState(null, document.title, tab.attr('href'));
                    }
                    return undefined;
                },
                dataType: 'script'
            })
        }
        else {
            if ("replaceState" in window.history) {
                window.history.replaceState(null, document.title, tab.attr('href'));
            }
        }
    })
};

function selectAllOptions(id) {
  var select = $('#'+id);
  select.children('option').attr('selected', true);
}


function submit_query_form(id) {
  selectAllOptions("selected_columns");
  $('#'+id).submit();
}

//replaces redmine default method showTab() beacuse of compatibility Redmine 3.1+
function showContactTab(name, url) {
  $('div#content .tab-content').hide();
  $('div.tabs a').removeClass('selected');
  $('#tab-content-' + name).show();
  $('#tab-' + name).addClass('selected');
  if ("replaceState" in window.history) {
    window.history.replaceState(null, document.title, url);
  }
  return false;
}

function tooglePriceField() {
  $('#deal_price').prop( "disabled", $('.line:visible').size() > 0 );
}

function toggleCRMIssuesSelection(el) {
  var boxes = $(el).parents('form').find('input[type=checkbox]');
  var all_checked = true;
  boxes.each(function(){ if (!$(this).prop('checked')) { all_checked = false; } });
  boxes.each(function(){
    if (all_checked) {
      $(this).removeAttr('checked');
      $(this).parents('tr').removeClass('context-menu-selection');
    } else if (!$(this).prop('checked')) {
      $(this).prop('checked', true);
      $(this).parents('tr').addClass('context-menu-selection');
    }
  });
}

function toogleDealItems(el) {
  $(el).closest('p').hide();
  $('.deal_items').show();
}

function uploadAvatar(element) {
  var fileSpan = $('<span>', { id: 'attachments_0'});
  $('#contact_data #attachments_fields').html('');
  fileSpan.append(
      $('<input>', { type: 'hidden', class: 'filename', name: 'attachments[0][filename]'} ).val(element.files[0].name),
      $('<input>', { type: 'hidden', class: 'description', name: 'attachments[0][description]' } ).val('avatar'),
      $('<input>', { type: 'hidden', class: 'token', name: 'attachments[0][token]' } )
  ).appendTo('#contact_data #attachments_fields');
  ajaxUpload(element.files[0], 0, fileSpan, element);
  return 0;
}


function initDealSelect2(id, url, placeholder) {
    $(function () {
        $('select#' + id).select2({
            ajax: {
                url: url,
                dataType: 'json',
                delay: 250,
                data: function (params) {
                    return { q: params.term };
                },
                processResults: function (data, params) {
                    return { results: data };
                },
                cache: true
            },
            placeholder: placeholder,
            allowClear: true,
            minimumInputLength: 0,
            width: '60%',
            templateResult: function (option) {
                return $('<span>' + option.avatar + '&nbsp;' + option.text + '</span>');
            }
        });
    });
};
